kernel/src/table_changes/schema_compat.rs

```diff
@@ -0,0 +1,302 @@
+use std::collections::{HashMap, HashSet};
+
+use crate::schema::{DataType, Schema, StructField, StructType};
+
+fn is_nullability_compatible(existing_nullable: bool, read_nullable: bool) -> bool {
+    // The case to avoid is when the read_schema is non-nullable and the existing one is nullable.
+    // So we avoid the case where !read_nullable && existing_nullable
+    // Hence we check that !(!read_nullable && existing_nullable)
+    // == read_nullable || !existing_nullable
+    read_nullable || !existing_nullable
+}
+fn is_struct_read_compatible(existing: &StructType, read_type: &StructType) -> bool {
+    // Delta tables do not allow fields that differ in name only by case
+    let existing_fields: HashMap<&String, &StructField> = existing
+        .fields()
+        .map(|field| (field.name(), field))
+        .collect();
+
+    let existing_names: HashSet<String> = existing
+        .fields()
+        .map(|field| field.name().clone())
+        .collect();
+    let read_names: HashSet<String> = read_type
+        .fields()
+        .map(|field| field.name().clone())
+        .collect();
+    if !existing_names.is_subset(&read_names) {
+        return false;
+    }
+    read_type
+        .fields()
+        .all(|read_field| match existing_fields.get(read_field.name()) {
```

Comment:
It seems like we should only need to materialize a hash set for one side (build), and just stream the other side's fields past it (probe)?

Also: kernel's `StructType::fields` member is already an `IndexMap` so you should have O(1) name lookups without building any additional hash sets.